tests/test_utils/test_stations_utils.py

```diff
@@ -190,3 +126,139 @@ def test_00_location_code(self, df_from_inv):
+                    if seed in expected_missing and chan.response is not None:
+                        return False
+                    if seed not in expected_missing and chan.response is None:
+                        return False
+        return True
+
+    @pytest.fixture
+    def df_with_nrl_response(self):
+        """
+        Add NRL response information to the dataframe.
+
+        Note: The datalogger is probably not correct for the test station.
+        It is just to ensure the NRL can be used to get a response.
+        """
+        _inv = obsplus.load_dataset("bingham_test").station_client.get_stations()
+        inv = _inv.select(station="NOQ")
+
+        with suppress_warnings():
+            df = obsplus.stations_to_df(inv)
+
+        # set instrument str
+        sensor_keys = ("Nanometrics", "Trillium 120 Horizon")
+        # get digitizer keys
+        datalogger_keys = (
+            "Nanometrics",
+            "Centaur",
+            "1 Vpp (40)",
+            "Off",
+            "Linear phase",
+            "100",
+        )
+        # keep one as a tuple and convert the other to str
+        df["sensor_keys"] = [sensor_keys for _ in range(len(df))]
+        df["datalogger_keys"] = "__".join(datalogger_keys)
+        # drop seed id
+        df = df.drop(columns="seed_id")
+        return df
+
+    @pytest.fixture
+    def df_with_get_stations_kwargs(self):
+        """
+        Add response information to the dataframe using get_stations_kwargs.
+        """
+        _inv = obsplus.load_dataset("bingham_test").station_client.get_stations()
+        inv = _inv.select(station="NOQ")
+
+        with suppress_warnings():
+            df = obsplus.stations_to_df(inv).reset_index()
+
+        # set get_station_kwargs for last two channels, leave first empty
+        kwargs_list = [""]
+        for ind, row in df.iloc[1:].iterrows():
+            kwargs = {x: row[x] for x in NSLC}
+            kwargs["endafter"] = str(to_utc(row["start_date"]))
+            kwargs_list.append(kwargs)
+        # set last kwargs to str to simulate reading from csv
+        kwargs_list[-1] = str(kwargs_list[-1])
+        df["get_station_kwargs"] = kwargs_list
+        # set the first kwargs to a string to make sure it can be parsed
+        # this is important for eg reading data from a csv.
+        df.loc[0, "get_station_kwargs"] = str(df.loc[0, "get_station_kwargs"])
+        return df
+
+    @pytest.fixture
+    def df_with_partial_responses(self, df_with_nrl_response):
+        """ test creating inv with partial responses. """
+        # set one row to None
+        df_with_nrl_response.loc[0, "sensor_keys"] = None
+        return df_with_nrl_response
+
+    @pytest.fixture
+    def df_with_both_response_cols(
+        self, df_with_nrl_response, df_with_get_stations_kwargs
+    ):
+        """
+        Get a df with both types of responses.
+
+        row 1 should have NRL, row 2 both, and row three client kwargs
```

Comment:
Am I understanding correctly that row 2 has both the information to fetch the nominal response and to retrieve from a provided client? If so, how does one know which set of arguments take precedence?

Reply:
I added a note about this in the docs, I think the thing to do is just give the NRL precedence.

Reply:
All right. I don't have a strong opinion on this, but one thing to consider: should the station client actually take precedence over the NRL? That could theoretically have more specific information associated with it. Ideally, only one or the other would be specified for each station/channel and this doesn't really matter, though.